tracer/src/Datadog.Trace/Activity/OtlpHelpers.cs

```diff
@@ -0,0 +1,379 @@
+            // **** To reliably get this, we might consider addding a Processor to the TracerProviderBuilder, though not sure where to invoke it
+            // - service.instance.id
+            // - service.name
+            // - service.namespace
+            // - service.version
+
+            if (w3cActivity is not null)
+            {
+                span.SetTag("otel.trace_id", w3cActivity.TraceId);
+
+                // Marshall events here using tag "events"
+                // span.SetTag("events", eventsArray);
+            }
+
+            // Fixup "version" tag
+            if (Tracer.Instance.Settings.ServiceVersion is null
+                && span.GetTag("service.version") is string otelServiceVersion
+                && !string.IsNullOrEmpty(otelServiceVersion))
+            {
+                span.SetTag(Tags.Version, otelServiceVersion);
+            }
+
+            // Copy over tags from Activity to the Datadog Span
+            // Starting with .NET 5, Activity can hold tags whose value have type object?
+            // For runtimes older than .NET 5, Activity can only hold tags whose values have type string
+            if (activity5 is not null)
+            {
+                foreach (var activityTag in activity5.TagObjects)
+                {
+                    OtlpHelpers.SetTagObject(span, activityTag.Key, activityTag.Value);
+                }
+            }
+            else
+            {
+                foreach (var activityTag in activity.Tags)
+                {
+                    OtlpHelpers.SetTagObject(span, activityTag.Key, activityTag.Value);
+                }
+            }
+
+            // TODO: Add container tags from attributes if the tag isn't already in the span
+
+            // Fixup "env" tag
+            if (span.Context.TraceContext?.Environment is null
+                && span.GetTag("deployment.environment") is string otelServiceEnv
+                && !string.IsNullOrEmpty(otelServiceEnv))
+            {
+                span.SetTag(Tags.Env, otelServiceEnv);
+            }
+
+            // TODO: The .NET OTLP exporter doesn't currently add tracestate, but the DD agent will set it as tag "w3c.tracestate" if detected
+            // For now, we can keep this unimplemented so the resulting span matches the .NET OTLP exporter
+            // span.SetTag("w3c.tracestate", w3CActivity.TraceStateString);
+
+            // Add the library name and library version
+            if (activity5 is not null)
+            {
+                span.SetTag("otel.library.name", activity5.Source.Name);
+                span.SetTag("otel.library.version", activity5.Source.Version);
+            }
+
+            // Set OTEL status code and OTEL status description
+            if (span.GetTag("otel.status_code") is null)
+            {
+                span.SetTag("otel.status_code", "STATUS_CODE_UNSET");
+            }
```

Comment:
As these are _always_ added, maybe it's worth having a special OtelTags and special casing here? e.g. 

```csharp
if(span is otelTags)
{
    if (activity5 is not null)
    {
        otelTags.LibraryName = activity5.Source.Name;
        otelTags.LibraryVersion = activity5.Source.Version;
    }
    // can default otelTags.StatusCode = "STATUS_CODE_UNSET" in OtelTags object itself.
}
else
{     
    if (activity5 is not null)
    {
        span.SetTag("otel.library.name", activity5.Source.Name);
        span.SetTag("otel.library.version", activity5.Source.Version);
    }
  
    if (span.GetTag("otel.status_code") is null)
    {
        span.SetTag("otel.status_code", "STATUS_CODE_UNSET");
    }

}
```

Reply:
I like this refactoring, though maybe as a follow-up PR